### PR TITLE
CodeKart models, associations and tests

### PR DESCRIPTION
**Pull request by carlospazuzu**
## Description
This PR uploads all files needed to build the Codekart database. 

## How was this tested?
`RSpec 6.0.0` was used to build the model tests of this PR. It was also used `shoulda matches` gem and `factory bot rails` for factories helpers.

## Database Diagram
image

### Racer
- ### Fields
- - - [x] `name`: Pilot's name. Obligatory
- - - [x] `born_at`: Pilot's birth date. Obligatory. 
- - - - [x] Pilot must be at least 18 years old in order to be elegible to participate a race
- - - [x] `img_url`: Pilot's image URL. Optional.
- ### Associations
- - - [x] A racer can participate many racers
- - - [x] A racer can have only one placement per race

### Placement
- ### Fields
- - - [x] `position`: Placement of a pilot in the current race
- - - - [x] It shouldn't have more than a player occuping the same placement in the same race
- ### Associations
- - - [x] A placement belongs to a racer
- - - [x] A placement belongs to a race

### Race
- ### Fields
- - - [x] `date`: Date that occuried the race. Obligatory
- - - [x] `place`: Place where the race happened. Obligatory
- ### Associations
- - - [x] A race has many racers
- - - [x] A race has only one placement per racer
- - - [x] A race belongs to a Tournament

### Tournament
- ### Fields
- - - [x] `name`: Name of the Tournament. Obligatory
- ### Associations
- - - [x] A Tournament has many races

### CODE REVIEW

File: app/models/race.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+  has_many :placements, dependent: :destroy
+  has_many :racers, through: :placements, dependent: :destroy
+
+  accepts_nested_attributes_for :placements
+
+  belongs_to :tournament
+
+  def add_racer_placement(racer_id, position)
+    Placement.create({position: position, racer_id: racer_id, race_id: self.id})
+  end
```
Comment (tonyaraujop) [inline app/models/race.rb]: Please remove this method. Is only used for a test case that can be rewritten.

File: spec/models/placement_spec.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    context 'when trying to attribute the same position to two different racers' do
+      it 'rolls back the conflicting operation and notify that a position cannot be occupied by more than one racer' do
+        race = create(:race)
+        racer1 = create(:racer)
+        racer2 = create(:racer)
+        Placement.create({ race_id: race.id, racer_id: racer1.id, position: 1 })
+        result = Placement.create({ race_id: race.id, racer_id: racer2.id, position: 1 })
+        expect(result.errors[:position].first).to eq('A position cannot be occupied for more than one racer!')
+      end
+    end
```
Comment (tonyaraujop) [inline spec/models/placement_spec.rb]: Both context can be written using shoulda-matcher. Take a look at this please https://matchers.shoulda.io/docs/v5.3.0/Shoulda/Matchers/ActiveRecord.html#validate_uniqueness_of-instance_method

File: spec/models/racer_spec.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+      it 'generates an error object in the errors array' do
+        underage_racer.valid?
+        expect(underage_racer.errors[:born_at].first).to eq('Need to be at least 18 years old!')
+      end
+    end
+  end
+
+  describe 'Associations' do
+    it { is_expected.to have_many(:races) }
+  end
```
Comment (tonyaraujop) [inline spec/models/racer_spec.rb]: A racer also have association with placement. Please, make a test for it too.

File: spec/models/racer_spec.rb [hunk 4]
```diff
@@ -0,0 +1,25 @@
+require 'rails_helper'
+
+RSpec.describe Racer, type: :model do
+  describe 'Validations' do
+    it { is_expected.to validate_presence_of(:name)    }
+    it { is_expected.to validate_presence_of(:born_at) }
+
+    context 'when trying to register an underaged racer' do
```
Comment (tonyaraujop) [inline spec/models/racer_spec.rb]:
```suggestion
    context 'when racer is underaged (18 years old)' do
```

File: spec/models/racer_spec.rb [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+      let(:underage_racer) { build(:racer, :underaged) }
+
+      it 'marks the object as invalid' do
```
Comment (tonyaraujop) [inline spec/models/racer_spec.rb]:
```suggestion
      it 'racer is invalid' do
```

File: spec/models/racer_spec.rb [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+        expect(underage_racer.invalid?).to be true
+      end
+
+      it 'generates an error object in the errors array' do
```
Comment (tonyaraujop) [inline spec/models/racer_spec.rb]:
```suggestion
      it 'returns an error' do
```

File: spec/models/racer_spec.rb (hunk 4 above)
Comment (tonyaraujop) [inline spec/models/racer_spec.rb]: You also need to make a context for the valid case (when pilot isn't underaged).

**Review by tonyaraujop — CHANGES_REQUESTED** [review]
Great job so far. But we can make it better.

File: app/models/race.rb (hunk 1 above)
Comment (carlospazuzu) [reply app/models/race.rb]: Indeed! Removed!

File: spec/models/racer_spec.rb (hunk 3 above)
Comment (carlospazuzu) [reply spec/models/racer_spec.rb]: Added in the next commit!

**Review by tonyaraujop — APPROVED** [review]
LGTM

**Pull request merged**